hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestSparkSql.scala

```diff
@@ -0,0 +1,518 @@
+    val fs = FSUtils.getFs(basePath, spark.sparkContext.hadoopConfiguration)
+
+    //Insert Operation
+    val dataGen = new HoodieTestDataGenerator(HoodieTestDataGenerator.TRIP_NESTED_EXAMPLE_SCHEMA, 0xDEED)
+    val (inputDf0, _, values0) = generateInserts(dataGen, "000", 100)
+    insertInto(tableName, values0, "bulk_insert", isMetadataEnabledOnWrite, keyGenClass)
+    assertTrue(HoodieDataSourceHelpers.hasNewCommits(fs, basePath, "000"))
+
+    //Snapshot query
+    val snapshotDf1 = doRead(tableName, isMetadataEnabledOnRead)
+    snapshotDf1.cache()
+    assertEquals(100, snapshotDf1.count())
+
+    val (updateDf, _, values1) = generateUniqueUpdates(dataGen, "001", 50)
+    insertInto(tableName, values1, "upsert", isMetadataEnabledOnWrite, keyGenClass)
+
+    val commitInstantTime2 = HoodieDataSourceHelpers.latestCommit(fs, basePath)
+
+    val snapshotDf2 = doRead(tableName, isMetadataEnabledOnRead)
+    snapshotDf2.cache()
+    assertEquals(100, snapshotDf2.count())
+
+    compareUpdateDfWithHudiDf(updateDf, snapshotDf2, snapshotDf1, colsToSelect)
+
+    val (inputDf2, _, values2) = generateUniqueUpdates(dataGen, "002", 60)
+    val uniqueKeyCnt2 = inputDf2.select("_row_key").distinct().count()
+    insertInto(tableName, values2, "upsert", isMetadataEnabledOnWrite, keyGenClass)
+
+    val commitInstantTime3 = HoodieDataSourceHelpers.latestCommit(fs, basePath)
+    assertEquals(3, HoodieDataSourceHelpers.listCommitsSince(fs, basePath, "000").size())
+
+    // Snapshot Query
+    val snapshotDf3 = doRead(tableName, isMetadataEnabledOnRead)
+    snapshotDf3.cache()
+    assertEquals(100, snapshotDf3.count())
+
+    compareUpdateDfWithHudiDf(inputDf2, snapshotDf3, snapshotDf3, colsToSelect)
+
+    // Read Incremental Query, need to use spark-ds because functionality does not exist for spark sql
+    // we have 2 commits, try pulling the first commit (which is not the latest)
+    val firstCommit = HoodieDataSourceHelpers.listCommitsSince(fs, basePath, "000").get(0)
+    val hoodieIncViewDf1 = spark.read.format("org.apache.hudi")
+      .option(DataSourceReadOptions.QUERY_TYPE.key, DataSourceReadOptions.QUERY_TYPE_INCREMENTAL_OPT_VAL)
+      .option(DataSourceReadOptions.BEGIN_INSTANTTIME.key, "000")
+      .option(DataSourceReadOptions.END_INSTANTTIME.key, firstCommit)
+      .load(basePath)
+    //val hoodieIncViewDf1 = doIncRead(tableName, isMetadataEnabledOnRead, "000", firstCommit)
+    assertEquals(100, hoodieIncViewDf1.count()) // 100 initial inserts must be pulled
+    var countsPerCommit = hoodieIncViewDf1.groupBy("_hoodie_commit_time").count().collect()
+    assertEquals(1, countsPerCommit.length)
+    assertEquals(firstCommit, countsPerCommit(0).get(0).toString)
+
+    val (inputDf3, _, values3) = generateUniqueUpdates(dataGen, "003", 80)
+    val uniqueKeyCnt3 = inputDf3.select("_row_key").distinct().count()
+    insertInto(tableName, values3, "upsert", isMetadataEnabledOnWrite, keyGenClass)
```

Comment:
good to have insertInto for all write operations. but would like to see MERGE INTO as well. bcoz, thats widely used as well.

Reply:
```
merge into hudi_mor_tbl as target
using merge_source as source
on target.id = source.id
when matched then update set *
when not matched then insert *
;
```